atl-checker/src/algorithms/global/multithread.rs

```diff
@@ -0,0 +1,484 @@
+                    self.assignment_mut().insert(task.clone(), false);
+                }
+                for edge in self.edg.succ(&task) {
+                    match edge {
+                        Edge::Hyper(e) => {
+                            self.process_hyper(e);
+                        }
+                        Edge::Negation(e) => {
+                            self.process_negation(e);
+                        }
+                    }
+                    trace!("Worker finished task");
+                }
+                self.broker
+                    .send_result(task.clone(), *self.assignment.get(&task).unwrap());
+            }
+        }
+    }
+}
+
+pub struct MultithreadedGlobalAlgorithm<G: ExtendedDependencyGraph<V>, V: Vertex> {
+    edg: G,
+    worker_count: u64,
+    v0: V,
+    assignment: HashMap<V, bool>,
+    dist: VecDeque<HashSet<V>>,
+}
+
+impl<
+        G: ExtendedDependencyGraph<V> + Send + Sync + Clone + Debug + 'static,
+        V: Vertex + Send + Sync + 'static,
+    > GlobalAlgorithm<G, V> for MultithreadedGlobalAlgorithm<G, V>
+{
+    fn edg(&self) -> &G {
+        &self.edg
+    }
+    fn edg_mut(&mut self) -> &mut G {
+        &mut self.edg
+    }
+    fn v0(&self) -> &V {
+        &self.v0
+    }
+    fn v0_mut(&mut self) -> &mut V {
+        &mut self.v0
+    }
+    fn assignment(&self) -> &HashMap<V, bool> {
+        &self.assignment
+    }
+    fn assignment_mut(&mut self) -> &mut HashMap<V, bool> {
+        &mut self.assignment
+    }
+    fn dist(&self) -> &VecDeque<HashSet<V>> {
+        &self.dist
+    }
+    fn dist_mut(&mut self) -> &mut VecDeque<HashSet<V>> {
+        &mut self.dist
+    }
+
+    fn run(&mut self) -> bool {
+        self.initialize();
+
+        let (mut brokers, manager) = GChannelBroker::new(self.worker_count);
+        for _ in 0..self.worker_count {
+            let mut worker = GWorker::new(
+                self.edg.clone(),
+                self.assignment.clone(),
+                brokers.pop().unwrap(),
+            );
+            thread::spawn(move || {
+                worker.run();
+            });
+        }
+
+        let components = self.dist.clone();
+        components.iter().rev().for_each(|component| {
+            let mut tasks: HashMap<V, bool> = HashMap::new();
+            let mut changed_flag = true;
+
+            for vertex in component {
+                self.queue_task(vertex.clone(), &mut tasks, &manager);
+            }
+
+            loop {
+                if !tasks.values().any(|value| *value) {
+                    if manager.task_queue_is_empty() && !changed_flag {
+                        break;
+                    }
+
+                    manager.send_updates(self.assignment().clone());
+                    for vertex in component {
+                        self.queue_task(vertex.clone(), &mut tasks, &manager);
+                    }
+                    changed_flag = false;
+                }
```

Comment:
Is it important to keep track of which tasks are still queued? If we just care about if any task is still running, then you could use an counter instead of a hashmap. Increase the counter whenever you queue a task, decrease when it is complete. It is faster to increase/decrease/check-if-0 than it is to compute the hashes of vertices over and over.